core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -143,9 +143,11 @@ class LogSegment(val log: FileMessageSet,
           // no max offset, just read until the max position
           min((maxPosition - startPosition.position).toInt, maxSize)
         case Some(offset) => {
-          // there is a max offset, translate it to a file position and use that to calculate the max read size
+          // there is a max offset, translate it to a file position and use that to calculate the max read size;
+          // if the max offset is smaller than the start offset, return empty response as it can happen when the max offset
+          // is bounded by the high watermark for a fetch request from normal consumer.
           if(offset < startOffset)
-            throw new IllegalArgumentException("Attempt to read with a maximum offset (%d) less than the start offset (%d).".format(offset, startOffset))
+            return FetchDataInfo(offsetMetadata, MessageSet.Empty)
```

Comment:
This is a bit subjective, but I would avoid the explicit `return` here and use a `-1` length as a marker for an empty response (which would be checked just before creating a `FetchDataInfo` at the end of the method). The code would be a bit more regular. But fine to leave as is if you prefer it this way. Also, if you do decide to change, please take the chance to move `maxOffset match {` to the previous line. It's creating an additional level of indenting for no reason.